### PR TITLE
[#IP-278] ServicePreferencesSettings version update when mode changes

### PR DESCRIPTION
**Pull request by michaeldisaro**
#### List of Changes
- ServicePreferencesSettings version update when mode changes during user update

#### Motivation and Context
We want to track a new version of settings if the user changes the mode of his service's preferences.

#### How Has This Been Tested?
It has been tested by performing:
- yarn install --frozen-lockfile
- yarn build
- yarn lint
- yarn test

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: UpdateProfile/handler.ts [hunk 1]
```diff
@@ -97,10 +99,35 @@ export function UpdateProfileHandler(
       profilePayload.email !== undefined &&
       profilePayload.email !== existingProfile.email;
 
+    // Check if servicePreferencesSettings mode changed
+    const servicePreferencesSettingsModeChanged =
+      profilePayload.service_preferences_settings !== undefined &&
+      profilePayload.service_preferences_settings.mode !==
+        existingProfile.servicePreferencesSettings.mode;
+
+    // check if user is sending LEGACY as mode
```
Comment (BurnedMarshal) [inline UpdateProfile/handler.ts:108]: Can we change this comment with a more clear description? for es.
```
// returns to LEGACY profile from updated ones is forbidden 
```

File: openapi/index.yaml [hunk 2]
```diff
@@ -416,85 +416,91 @@ parameters:
     x-example: DOWNLOAD
 definitions:
   ExtendedProfile:
-    $ref: "https://raw.githubusercontent.com/pagopa/io-functions-commons/v20.4.0/openapi/definitions.yaml#/ExtendedProfile"
+    $ref: "https://raw.githubusercontent.com/pagopa/io-functions-commons/v20.6.1/openapi/definitions.yaml#/ExtendedProfile"
```
Comment (BurnedMarshal) [inline openapi/index.yaml:419]: Pls upgrade all `io-functions-commons` references to `20.6.5` specification

Comment (michaeldisaro) [inline openapi/index.yaml:419]: Shouldn't all these be v20.6.5?

Comment (michaeldisaro) [reply openapi/index.yaml:419]: @balanza @BurnedMarshal

File: utils/profiles.ts [hunk 3]
```diff
@@ -30,7 +32,18 @@ export function apiProfileToProfile(
     isEmailValidated,
     isInboxEnabled: apiProfile.is_inbox_enabled,
     isWebhookEnabled: apiProfile.is_webhook_enabled,
-    preferredLanguages: apiProfile.preferred_languages
+    preferredLanguages: apiProfile.preferred_languages,
+    servicePreferencesSettings:
+      apiProfile.service_preferences_settings.mode ===
```
Comment (BurnedMarshal) [inline utils/profiles.ts:37]: What if `service_preferences_settings` is `undefined`?

File: UpdateProfile/__tests__/handler.test.ts [hunk 4]
```diff
@@ -109,6 +115,223 @@ describe("UpdateProfileHandler", () => {
...
+    };
+
+    const updateProfileHandler = UpdateProfileHandler(profileModelMock as any);
+
+    const result = await updateProfileHandler(contextMock as any, aFiscalCode, {
+      ...aProfile,
+      service_preferences_settings: legacyApiProfileServicePreferencesSettings
+    });
+
+    expect(result.kind).toBe("IResponseErrorConflict");
```
Comment (BurnedMarshal) [inline UpdateProfile/__tests__/handler.test.ts:134]:
```suggestion
    expect(result.kind).toBe("IResponseErrorConflict");
    expect(profileModelMock.findLastVersionByModelId).toBeCalled();
    expect(profileModelMock.update).not.toBeCalled();
```

File: UpdateProfile/__tests__/handler.test.ts [hunk 5]
```diff
@@ -109,6 +115,223 @@ describe("UpdateProfileHandler", () => {
...
```
Comment (BurnedMarshal) [inline UpdateProfile/__tests__/handler.test.ts:153]: the same in here
```suggestion
    expect(result.kind).toBe("IResponseErrorConflict");
    expect(profileModelMock.findLastVersionByModelId).toBeCalled();
    expect(profileModelMock.update).not.toBeCalled();
```

File: openapi/index.yaml (hunk 2 above)
Comment (gquadrati) [reply openapi/index.yaml:419]: @michaeldisaro @balanza  in my PR #170 I've switched to openapi generation from index.openapi.template.
In this way we removed this reference by importing deinitions from fn-commons node_module (see here: https://github.com/pagopa/io-functions-app/blob/ccd26f65dfaa2a6e7f73bbcdc7b615388549697a/openapi/index.yaml.template#L458)

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: openapi/index.yaml (hunk 2 above)
Comment (michaeldisaro) [reply openapi/index.yaml:419]: Seems good. So we don't have to keep it updated anymore.

File: UpdateProfile/handler.ts (hunk 1 above)
Comment (michaeldisaro) [reply UpdateProfile/handler.ts:108]: Ok, seems legit.

File: utils/profiles.ts (hunk 3 above)
Comment (michaeldisaro) [reply utils/profiles.ts:37]: We keep the old one, be it the default or another one. Does it seem legit?

File: openapi/index.yaml (hunk 2 above)
Comment (michaeldisaro) [reply openapi/index.yaml:419]: @gquadrati you did this, as said above, right?

Comment (gquadrati) [reply openapi/index.yaml:419]: Yes, exactly

File: utils/profiles.ts [hunk 6]
```diff
@@ -39,11 +40,11 @@ export function apiProfileToProfile(
         ServicesPreferencesModeEnum.LEGACY
         ? {
             mode: ServicesPreferencesModeEnum.LEGACY,
-            version: 0
+            version: -1
```
[the remark this answers is not in the document]
Comment (michaeldisaro) [reply utils/profiles.ts:43]: Ok, I fix it.

[the remark this answers is not in the document]
Comment (michaeldisaro) [reply utils/profiles.ts]: Resolved in 5e32066

File: UpdateProfile/__tests__/handler.test.ts [hunk 7]
```diff
@@ -109,6 +115,277 @@ describe("UpdateProfileHandler", () => {
...
+          service_preferences_settings: {
+            mode: manualProfileServicePreferencesSettings.mode,
+            version: manualProfileServicePreferencesSettings.version
+          }
+        })
+      );
+    }
+  });
+
+  it("should increment service_preferences_settings.version if mode has changed from LEGACY to MANUAL", async () => {
```
Comment (BurnedMarshal) [inline UpdateProfile/__tests__/handler.test.ts:296]: is missing the test case where we change the mode from LEGACY to AUTO.

File: UpdateProfile/handler.ts [hunk 8]
```diff
@@ -97,10 +99,36 @@ export function UpdateProfileHandler(
       profilePayload.email !== undefined &&
       profilePayload.email !== existingProfile.email;
 
+
+    // Get servicePreferencesSettings mode from payload or default to LEGACY
+    // tslint:disable-next-line: prettier
+    const requestedServicePreferencesSettingsMode = profilePayload.service_preferences_settings?.mode ?? ServicesPreferencesModeEnum.LEGACY;
```
Comment (BurnedMarshal) [inline UpdateProfile/handler.ts:105]: What if?
```suggestion
    const requestedServicePreferencesSettingsMode = fromNullable(
      profilePayload.service_preferences_settings
    )
      .map(_ => _.mode)
      .getOrElse(ServicesPreferencesModeEnum.LEGACY);
```

File: UpdateProfile/handler.ts [hunk 9]
```diff
@@ -97,10 +99,36 @@ export function UpdateProfileHandler(
...
+
+    // Check if a mode change is requested
+    const servicePreferencesSettingsModeChanged = requestedServicePreferencesSettingsMode !== existingProfile.servicePreferencesSettings.mode;
```
Comment (BurnedMarshal) [inline UpdateProfile/handler.ts:108]: Can we use `is` prefix for constants and functions that are or return boolean values?
```suggestion
    const isServicePreferencesSettingsModeChanged = requestedServicePreferencesSettingsMode !== existingProfile.servicePreferencesSettings.mode;
```

File: UpdateProfile/__tests__/handler.test.ts (hunk 7 above)
Comment (michaeldisaro) [reply UpdateProfile/__tests__/handler.test.ts:317]: fixed in 01360d9

File: UpdateProfile/handler.ts (hunk 8 above)
Comment (michaeldisaro) [reply UpdateProfile/handler.ts]: fixed in 01360d9

File: UpdateProfile/handler.ts (hunk 9 above)
Comment (michaeldisaro) [reply UpdateProfile/handler.ts]: fixed in 01360d9

**Review by BurnedMarshal — APPROVED** [review]
LGTM

**Pull request merged**